### PR TITLE
Add Toil appliance (resolves #1142)

### LINKED ISSUE #1142: Toil build should create and optionally push Docker appliance
Avoids chicken-eggyness of current approach where appliance is built in cgl-docker-lib and allows for building developer appliances from a dirty Toil working copy, including running the IT with those developer appliances.

### PR DESCRIPTION
**Pull request by hannes-ucsc**
Also changes some Makefile mechanics (printf instead of echo, switch to bash as subshell). Thanks to @cket and @fnothaft for writing the initial version of the Dockerfile.

Connected to #1088.

Also sneaks in a change that moves the `retry` context manager to bd2k-python-lib.

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
@cket and @fnothaft, this is getting close to being ready. This would be a good time to review it.

File: Makefile [hunk 1]
```diff
@@ -64,32 +78,72 @@ clean_develop: check_venv
...
+docker: check_docker_registry docker/worker/Dockerfile docker/leader/Dockerfile
+	@set -ex \
+	; cd docker \
+	; for role in leader worker; do \
+	    docker build --tag=$(docker_registry)/$(docker_base_name)-$${role}:$(docker_tag) \
+	                 -f $${role}/Dockerfile \
+	                 . \
+	; done
+	docker tag -f $(docker_registry)/$(docker_base_name)-leader:$(docker_tag) \
+	              $(docker_registry)/$(docker_base_name):$(docker_tag)
```
Comment (fnothaft) [inline Makefile]: Shouldn't this be `$(docker_registry)/$(docker_base_name)-worker:$(docker_tag)`?

Comment (fnothaft) [inline Makefile]: Er, nvm.

File: Makefile [hunk 2]
```diff
@@ -64,32 +78,72 @@ clean_develop: check_venv
...
+
+docker/%/Dockerfile: docker/Dockerfile.py docker/$(sdist_name)
+	mkdir -p docker/$*
+	$(python) docker/Dockerfile.py \
+	    --role=$* \
+	    --sdist=$(sdist_name) \
+	    --self=$(docker_registry)/$(docker_base_name)-$*:$(docker_tag) > $@
+
+clean_docker:
+	-rm docker/Dockerfile.{leader,worker} docker/$(sdist_name)
```
Comment (fnothaft) [inline Makefile]: Perhaps `docker rmi` on the built images as well?

File: docker/Dockerfile.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    COPY {options.sdist} .
+    RUN pip install {options.sdist}[aws,mesos,encryption,cwl]
+
+    # Move the Mesos module onto the python path
+    RUN ln -s /usr/lib/python2.7/site-packages/mesos /usr/local/lib/python2.7/dist-packages/mesos
+
+    # Fix for https://issues.apache.org/jira/browse/MESOS-3793
+    ENV MESOS_LAUNCHER=posix
+
+    # Embed a reference to the image itself so the container can do things like launch another
```
Comment (fnothaft) [inline docker/Dockerfile.py]: Small nit, but this comment reads a bit wonky. I had to read it several times before I understood what it was saying.

File: src/toil/jobStores/utils.py [hunk 4]
```diff
@@ -1,145 +1,12 @@
...
 from abc import abstractmethod
-from contextlib import contextmanager
-
-import time
 
 from bd2k.util.threading import ExceptionalThread
 
 log = logging.getLogger(__name__)
 
-
```
Comment (fnothaft) [inline src/toil/jobStores/utils.py]: Why are these changes coming in this PR?

**Comment by fnothaft** [comment]
LGTM overall, dropped a few comments inline.

**Pull request merged**